game/src/main/kotlin/world/gregs/voidps/world/interact/entity/npc/combat/type/GiantMole.kts

```diff
@@ -0,0 +1,209 @@
+}
+
+interfaceOption(component = "stayout", id = "warning_dark") {
+    player.closeInterfaces()
+}
+
+combatAttack {
+    val npc = target as NPC
+    if(npc.id == "giant_mole") {
+        val currentHealth = npc.levels.get(Skill.Constitution)
+        var shouldBurrow = false
+        if (it.fightStyle == "magic" && damage != 0) {
+            shouldBurrow = shouldBurrowAway(currentHealth)
+        } else if (it.fightStyle != "magic") {
+            shouldBurrow = shouldBurrowAway(currentHealth)
+        }
+        if(shouldBurrow && !World.timers.contains("await_mole_burrowing")) {
+            giantMoleBurrow(npc)
+        }
+    }
+}
+
+fun giantMoleBurrow(mole: NPC) {
+    mole.attackers.clear()
+    var tileToDust = getTotalDirection(mole.facing, mole.tile)
+    World.queue("await_mole_to_face", 1) {
+        if (tileToDust == Tile.EMPTY) {
+            logger.info { "failed to get facing tile for Giant Mole, using default tile." }
+            tileToDust = initialCaveTile
+        }
+        mole.face(tileToDust)
+        World.queue("display_burrow_dust", 1) {
+            if (shouldThrowDirt()) {
+                handleDirtOnScreen(mole.tile)
+            }
+            mole.setAnimation("giant_mole_burrow")
+            areaSound("giant_mole_burrow_down", mole.tile)
+            areaGraphic("burrow_dust", tileToDust)
+            World.queue("await_mole_burrowing", 1) {
+                val newLocation = gianMoleSpawns.random(mole)
+                mole.tele(newLocation!!)
+                mole.setAnimation("mole_burrow_up")
+            }
+        }
+    }
+}
+
+fun getTotalDirection(facing: Direction, moleTile: Tile): Tile {
+    return when (facing) {
+        Direction.NORTH -> Tile(moleTile.x, moleTile.y + 1)
+        Direction.SOUTH -> Tile(moleTile.x, moleTile.y - 1)
+        Direction.EAST -> Tile(moleTile.x + 1, moleTile.y)
+        Direction.WEST -> Tile(moleTile.x - 1, moleTile.y)
+        Direction.NORTH_EAST, Direction.SOUTH_EAST -> Tile(moleTile.x + 1, moleTile.y)
+        Direction.NORTH_WEST, Direction.SOUTH_WEST -> Tile(moleTile.x - 1, moleTile.y)
+        Direction.NONE -> Tile.EMPTY
+    }
+}
+
+// 13% chance to throw dirt on players screen
+fun shouldThrowDirt(): Boolean {
+    val dirtChance = Random.nextInt(0, 100)
+    return dirtChance <= 13
+}
+
+fun handleDirtOnScreen(moleTile: Tile) {
+    val nearMole = mutableListOf<Player>()
+    for (tile in moleTile.toCuboid(5)) {
+        for (player in players[tile]) {
+            nearMole.add(player)
+        }
+    }
+    for (player in nearMole) {
+        player.open("dirt_on_screen")
+        val playerInventory = player.inventory.items
+        for (item in playerInventory) {
+            if (item.id.contains("candle_lit")) {
+                val newItem = item.id.replace("_lit", "")
+                player.inventory.transaction {
+                    replace(item.id, newItem)
+                }
+            }
+        }
```

Comment:
```suggestion
        player.inventory.transaction {
             for (index in inventory.indices) {
                 val item = inventory[index]
                 if (item.id.endsWith("candle_lit")) {
                     replace(item.id, item.id.removeSuffix("_lit"))
                 }
             }
        }
```
Modifying the inventory inside loop for items could cause a concurrent modification exception